sched/scheduler/task_runner.go

```diff
@@ -78,32 +81,70 @@ func (r *taskRunner) run() error {
 	return err
 }
 
+// Run cmd and if there's a thrift error try re-running/re-querying until completion, retry timeout, or cmd timeout.
 func (r *taskRunner) runAndWait(taskId string, task sched.TaskDefinition) (runner.RunStatus, error) {
 	cmd := task.Command
 	if cmd.Timeout == 0 {
 		cmd.Timeout = r.defaultTaskTimeout
 	}
-
-	endTime := time.Now().Add(cmd.Timeout).Add(r.runnerOverhead)
-
-	st, err := r.runner.Run(&cmd)
-	if err != nil || st.State.IsDone() {
-		return st, err
+	cmdEndTime := time.Now().Add(cmd.Timeout).Add(r.runnerOverhead)
+	elapsedRetryTime := time.Duration(0)
+	var st runner.RunStatus
+	var err error
+	var id runner.RunID
+
+	// If thrift call returns an error then we treat it is a thrift error and will repeatedly retry.
+	// If thrift call returns a domain error that wasn't handled by the runner we should fail/return.
+	//TODO(jschiller): add a Nonce to Cmd so worker knows what to do if it sees a dup command?
+	log.Infof("Run() for job:%s taskId:%s", r.jobId, taskId)
+	for {
+		st, err = r.runner.Run(&cmd)
+		if err != nil && elapsedRetryTime+DefaultRetryInterval < r.thriftRetryTimeout {
+			log.Infof("Retrying run() for job:%s taskId:%s", r.jobId, taskId)
+			time.Sleep(DefaultRetryInterval)
+			elapsedRetryTime += DefaultRetryInterval
+			continue
+		} else if err != nil || st.State.IsDone() {
+			return st, err
+		}
+		break
 	}
+	id = st.RunID
 
-	id := st.RunID
-
+	log.Infof("Query(running) for job:%s, taskId:%s", r.jobId, taskId)
 	// Wait for the process to start running
-	st, err = r.queryWithTimeout(id, endTime, true)
-	if err != nil || st.State.IsDone() {
-		return st, err
+	elapsedRetryTime = 0
+	for {
+		st, err = r.queryWithTimeout(id, cmdEndTime, true)
+		if err != nil && elapsedRetryTime+DefaultRetryInterval < r.thriftRetryTimeout {
+			log.Infof("Retrying query(running) for job:%s, taskId:%s", r.jobId, taskId)
+			time.Sleep(DefaultRetryInterval)
+			elapsedRetryTime += DefaultRetryInterval
+			continue
+		} else if err != nil || st.State.IsDone() {
+			return st, err
+		}
+		break
 	}
 
 	// It's running, but not done, so we want to log a second StartTask that includes
```

Comment:
Suggest folding the 3rd loop into this one to get rid of duplication, and use a e.g. if notStarted flag + check for task log.  Also with this as-is, if the task exits quickly enough and the first poll in the 2nd loop gets IsDone, start task will never be logged?  Lastly would prefer for loop w/ conditionals vs manual control statements but willing to defer to preference